File: testng-core/src/test/java/test/attributes/issue2346/LocalTestListener.java
package test.attributes.issue2346;

import java.util.HashMap;
import java.util.Map;
import org.testng.ITestListener;
import org.testng.ITestResult;

public class LocalTestListener implements ITestListener {
  public static Map<String, Boolean> data = new HashMap<>();

  @Override
  public void onTestStart(ITestResult iTestResult) {
    String key = "onTestStart_" + iTestResult.getMethod().getQualifiedName();
    data.putIfAbsent(key, iTestResult.getAttributeNames().isEmpty());
  }

  @Override
  public void onTestSuccess(ITestResult iTestResult) {
    String key = "onTestSuccess_" + iTestResult.getMethod().getQualifiedName();
    data.putIfAbsent(key, iTestResult.getAttributeNames().isEmpty());
  }

  @Override
  public void onTestFailure(ITestResult iTestResult) {
    String key = "onTestFailure_" + iTestResult.getMethod().getQualifiedName();
    data.putIfAbsent(key, iTestResult.getAttributeNames().isEmpty());
  }

  @Override
  public void onTestSkipped(ITestResult iTestResult) {
    String key = "onTestSkipped_" + iTestResult.getMethod().getQualifiedName();
    data.putIfAbsent(key, iTestResult.getAttributeNames().isEmpty());
  }
}

File: testng-core/src/test/kotlin/test/attributes/samples/issue2346/LocalTestListener.kt
package test.attributes.samples.issue2346

import org.testng.ITestListener
import org.testng.ITestResult

val data = mutableMapOf<String, Boolean>()

class LocalTestListener : ITestListener {

    override fun onTestStart(result: ITestResult) {
        val key = "onTestStart_" + result.method.qualifiedName
        data.putIfAbsent(key, result.attributeNames.isEmpty())
    }

    override fun onTestSuccess(iTestResult: ITestResult) {
        val key = "onTestSuccess_" + iTestResult.method.qualifiedName
        data.putIfAbsent(key, iTestResult.attributeNames.isEmpty())
    }

    override fun onTestFailure(iTestResult: ITestResult) {
        val key = "onTestFailure_" + iTestResult.method.qualifiedName
        data.putIfAbsent(key, iTestResult.attributeNames.isEmpty())
    }

    override fun onTestSkipped(iTestResult: ITestResult) {
        val key = "onTestSkipped_" + iTestResult.method.qualifiedName
        data.putIfAbsent(key, iTestResult.attributeNames.isEmpty())
    }
}
